### PR TITLE
TS Refactor

### PR DESCRIPTION
**Pull request by agmoss**
Hi Travis!

I have undertaken an initial TS conversion of `random`. All code and tests are now in typescript with strict compiler options. My goal was to complete the refactor with minimal change to the patterns and flow of the existing code. Overall things have gone smoothly!

Things I could use an opinion on:

1. Build options
   - I would be happy to make a rollup config instead of a tsc build
2. There are a couple areas where I am using `any[]` to accommodate the unknown nature of some generics
3. Use of TSDoc vs JSDoc
4. Overhaul of rng-factory.ts to accommodate better type safety
5. Use of ow/ow-lite
  - ow lite does not have types. I can write these definitions or use ow or just use TS language features for validation

I still have some tinkering to do on this pr. I will have a more concrete deliverable in a few days. I would like some time to add to the test cases to further verify the validity of my code changes.

Cheers,

Andrew

### CODE REVIEW

**Comment by transitive-bullshit** [comment]
This is amazing work @agmoss! 🙏 

| I would be happy to make a rollup config instead of a tsc build

I prefer using `tsc` instead of rollup because it's less complicated and there's no real advantage to bundling your library code. Any modern bundler is going to have tree-shaking built-in, and if you bundle via rollup, that essentially gets nullified.

| There are a couple areas where I am using any[] to accommodate the unknown nature of some generics

Totally fine — let's get the TS version working first, and we can always improve the types over time.

| Use of TSDoc vs JSDoc

I'm not sure; I really like the current markdown docs, so if we can replicate that it'd be awesome, but I'm not sure if that's possible.

| Overhaul of rng-factory.ts to accommodate better type safety

Sounds good; again, I'd recommend getting an initial TS PR in first, and then we can always improve things over time.

| Use of ow/ow-lite

Not sure; the original reason I removed `ow` is that it made the package way too large. We might just want to remove it entirely, or add some lightweight types to ow-lite.

File: package.json [hunk 1]
```diff
@@ -63,8 +69,11 @@
     "dist"
   ],
   "ava": {
+    "extensions": [
+			"ts"
```
Comment (transitive-bullshit) [inline package.json:73]: we should add prettier in a follow-up PR (we could use this repo's usage of prettier as a template https://github.com/transitive-bullshit/react-static-tweets)

File: package.json [hunk 2]
```diff
@@ -43,13 +43,19 @@
...
+    "@types/seedrandom": "^2.4.28",
     "ava": "^3.5.0",
     "babel-cli": "^6.26.0",
     "babel-preset-env": "^1.7.0",
     "del-cli": "^1.1.0",
     "dev-cli": "^0.0.0",
     "esm": "^3.2.25",
     "standard": "^14.3.3",
+    "ts-node": "^9.1.1",
+    "tsc": "^1.20150623.0",
```
Comment (transitive-bullshit) [inline package.json:57]: shouldn't need to add `tsc` directly; it comes bundled with `typescript`.

File: package.json [hunk 3]
```diff
@@ -11,13 +11,13 @@
...
     "docs": "update-markdown-jsdoc -f src/random.js --shallow",
-    "build": "babel -s -d dist src",
+    "build": "tsc",
     "build:watch": "babel -w -s -d dist src",
     "prebuild": "yarn run clean",
     "clean": "del dist",
     "prepublishOnly": "yarn run build",
-    "pretest": "yarn run build",
-    "test": "ava -v && standard"
+    "test": "ava -v && standard",
```
Comment (transitive-bullshit) [inline package.json:19]: we may want to remove `standard` in favor of `eslint` w/ `typescript` ala https://github.com/transitive-bullshit/react-static-tweets

**Review by transitive-bullshit — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (agmoss) [reply package.json]: I can add prettier to this one if you would like

File: package.json (hunk 2 above)
Comment (agmoss) [reply package.json]: Noted

File: package.json [hunk 4]
```diff
@@ -11,13 +11,13 @@
   "scripts": {
     "start": "yarn build:watch",
     "docs": "update-markdown-jsdoc -f src/random.js --shallow",
-    "build": "babel -s -d dist src",
+    "build": "tsc",
     "build:watch": "babel -w -s -d dist src",
```
[the remark this answers is not in the document]
Comment (agmoss) [reply package.json]: Noted

File: package.json (hunk 3 above)
Comment (agmoss) [reply package.json]: Done

**Comment by agmoss** [comment]
Feedback has been incorporated. Tests have been expanded on to evaluate the custom `generators` (ie. xor128) , not just `seedrandom`. 

1. Using tsc
2. Anys identified with plan for removal
3. JS doc retained
   - This pr does not modify the original package api. Additionally I have not added to or changed the existing JSDoc comments.  Therefor a doc update is not really necessary. However updating the docs with the typescript change may pose an issue. As a workaround I pointed `update-markdown-jsdoc` to `dist/random.js`. This works but there are formatting and duplication implications. 
4. Leaving this for now 
5. Sticking with ow for now. If unpacked size is of concern this pr is ~same as original. Total size may be a different story. I was having trouble writing definitions on top of `ow-lite`. I could port `ow-lite` to ts faster than I could write independent type defs. If that is not acceptable then I would propose a removal of `ow-lite` in favor of some custom guards.

**Comment by agmoss** [comment]
- Created tests for invalid user inputs 
- Created tests for uniform int and uniform bool
- Expanded on test coverage for custom generators 
- Stronger type for RNG seed function

**Comment by transitive-bullshit** [comment]
Thanks so much @agmoss!

I merged your changes, made a few updates (added prettier & eslint, cleaned up a few things, updated readme), and just published a new major version `v3.0.0`.

I added you to the readme as well. 💯 

Thanks again && feel free to ping me if you wanna discuss what to work on next.

I'm guessing that we'll want to remove `ow` and probably `seedrandom` as well to reduce the bundle size as much as possible for browser usage.

**Comment by agmoss** [comment]
That is amazing @transitive-bullshit! Glad I could assist. Thanks for the shoutout! :boom:

**Pull request merged**